platforms/documentation/docs/src/docs/userguide/dep-man/declaring-depedencies/declaring_configurations.adoc

```diff
@@ -0,0 +1,148 @@
+- *Producer Role*: Configurations expose artifacts for consumption by other projects.
+
+=== Configurations for declaring dependencies
+
+To express that an application `app` _depends on_ library `lib`, _at least_ one configuration is required:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=declare-configuration]"]
+include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=declare-configuration]"]
+====
+
+*Configurations used for declaring dependencies* define and organize the dependencies required for different aspects of a project.
+
+=== Configurations for consumers (resolvable configuration)
+
+To specify how dependencies are used, you can create companion configurations:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=concrete-classpath]"]
+include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=concrete-classpath]"]
+====
+
+Now, `someConfiguration` declares dependencies, while `compileClasspath` and `runtimeClasspath` are resolvable configurations for specific purposes.
+When resolved they should contain the compile classpath, and the runtime classpath of the application respectively.
+
+*Resolvable configurations* are those that can be resolved to produce a set of files or artifacts.
+These configurations are used to define the classpath for different stages of a build process, such as compilation or runtime.
+
+=== Configurations for producers (consumable configuration)
+
+To specify what is exposed to consumers, you can create consumable configurations:
+
+====
+include::sample[dir="snippets/dependencyManagement/attributeMatching/kotlin",files="build.gradle.kts[tags=setup-configurations]"]
+include::sample[dir="snippets/dependencyManagement/attributeMatching/groovy",files="build.gradle[tags=setup-configurations]"]
+====
+
+Here, a library exposes configurations for its API and runtime.
+The `apiElements` configuration provides the API for compilation, while the `runtimeElements` configuration offers the runtime dependencies.
+These configurations are consumable but not resolvable, as indicated by the `canBeConsumed` and `canBeResolved` flags.
+
+*Consumable Configurations* expose artifacts for other projects to consume.
+They are not intended to be resolved directly but are meant to provide dependencies or resources that other projects can depend on.
+
+=== Configuration flags and roles
+
+Configurations have two key flags:
+
+- **`canBeResolved`:** Determines if a configuration can be resolved to a dependency graph. A configuration with `canBeResolved` set to `true` can be resolved, while one with `false` is only for declaring dependencies.
+- **`canBeConsumed`:** Indicates if a configuration is meant for consumers. A configuration with `canBeConsumed` set to `true` is exposed to other projects.
+
```

Comment:
We also have an incubating `isCanBeDeclared` which indicates whether you can declare dependencies on the configuration: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/Configuration.html#isCanBeDeclared()

In general, configurations should only have one of these flags enabled. 
A declarable configuration should not be resolvable or consumable.
A consumable configuration should not be declarable or resolvable.
A resolvable configuration should not be declarable or consumable.

We should probably add a section describing the incubating configuration factory methods on https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ConfigurationContainer.html (resolvable(), consumable(), dependencyScope()), which create configurations in these roles automatically, so users don't need to set the flags.

Reply:
Added but should be checked for accuracy.